src/dtgtk/range.h

```diff
@@ -0,0 +1,172 @@
+/*
+    This file is part of darktable,
+    Copyright (C) 2022 darktable developers.
+
+    darktable is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    darktable is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with darktable.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/* Precision about all the value reference used here
+  We have in fact 3 different reference
+  - the "real" value (for example ratio 0.67)
+    suffix = "_r"
+  - the "band" value : this is the value corrected in order to be shown on the graph.
+    this differ from value because for some property, we need non linear repartition of values
+    example : portrait ratio should be inverted
+    suffix = "_bd"
+  - the "pixels" value : this is the value in pixels as shown on the graph
+    this value changed each time the widget size change
+    suffix = "_px"
+*/
+#pragma once
+
+#include "common/datetime.h"
+#include "paint.h"
+#include <gtk/gtk.h>
+G_BEGIN_DECLS
+#define DTGTK_RANGE_SELECT(obj)                                                                                   \
+  G_TYPE_CHECK_INSTANCE_CAST(obj, dtgtk_range_select_get_type(), GtkDarktableRangeSelect)
+#define DTGTK_RANGE_SELECT_CLASS(klass)                                                                           \
+  G_TYPE_CHECK_CLASS_CAST(klass, dtgtk_range_select_get_type(), GtkDarktableRangeSelectClass)
+#define DTGTK_IS_RANGE_SELECT(obj) G_TYPE_CHECK_INSTANCE_TYPE(obj, dtgtk_range_select_get_type())
+#define DTGTK_IS_RANGE_SELECT_CLASS(klass) G_TYPE_CHECK_CLASS_TYPE(obj, dtgtk_range_select_get_type())
+
+typedef double (*DTGTKTranslateValueFunc)(const double value);
+typedef gchar *(*DTGTKPrintValueFunc)(const double value, const gboolean detailled);
+typedef gboolean (*DTGTKDecodeValueFunc)(const gchar *text, double *value);
+
+typedef enum dt_range_bounds_t
+{
+  DT_RANGE_BOUND_RANGE = 0,
+  DT_RANGE_BOUND_MIN = 1 << 0,
+  DT_RANGE_BOUND_MAX = 1 << 1,
+  DT_RANGE_BOUND_FIXED = 1 << 2,
+  DT_RANGE_BOUND_MAX_NOW = 1 << 3,
+  DT_RANGE_BOUND_MIN_RELATIVE = 1 << 4,
+  DT_RANGE_BOUND_MAX_RELATIVE = 1 << 5
+} dt_range_bounds_t;
+
+typedef enum dt_range_type_t
+{
+  DT_RANGE_TYPE_NUMERIC = 0,
+  DT_RANGE_TYPE_DATETIME
+} dt_range_type_t;
+
+typedef struct _GtkDarktableRangeSelect
+{
+  GtkBin widget;
+
+  dt_range_type_t type;
+
+  gboolean show_entries; // do we show the line with the entry boxes ?
+  double min_r;          // minimal value shown
```

Comment:
most of dt code is float, here (and in other part of this PR) you're using double. Is there a good reason to not use float?

Reply:
Well... I wanted to be absolutely sure that we don't get out of bounds for any reasons... As it's a "generic" widget which can potentially get a very different type of values, I've preferred to be safe... (but yes, for the cases we have actually, I think float will be enough)
Considering the quite small memory impact here, is there any reasons to prefer floats ?

Reply:
No, that was just a question. But indeed using double in a generic module is OK (that's what is used in cairo for example).